spanner/api/Spanner/Program.cs

```diff
@@ -1757,35 +1757,44 @@ await connection.BeginTransactionAsync())
-                new Singer {SingerId = firstSingerId, FirstName = "Marc",
-                    LastName = "Richards"},
-                new Singer {SingerId = secondSingerId, FirstName = "Catalina",
-                    LastName = "Smith"},
-                new Singer {SingerId = 3, FirstName = "Alice",
-                    LastName = "Trentor"},
-                new Singer {SingerId = 4, FirstName = "Lea",
-                    LastName = "Martin"},
-                new Singer {SingerId = 5, FirstName = "David",
-                    LastName = "Lomond"},
+            List<Album> albums = new List<Album> {
+                new Album { SingerId = secondSingerId, AlbumId = 1, AlbumTitle = "Green" },
+                new Album { SingerId = secondSingerId, AlbumId = 3, AlbumTitle = "Terrified" },
             };
             // Create connection to Cloud Spanner.
             using (var connection = new SpannerConnection(connectionString))
             {
                 await connection.OpenAsync();
 
-                // Delete rows from the Singers table.
-                var cmd = connection.CreateDeleteCommand("Singers",
+                // Delete individual rows from the Albums table.
+                var cmd = connection.CreateDeleteCommand(
+                    "Albums",
                     new SpannerParameterCollection {
-                        {"SingerId", SpannerDbType.Int64}
-                    });
-                await Task.WhenAll(singers.Select(singer =>
+                        {"SingerId", SpannerDbType.Int64},
+                        {"AlbumId", SpannerDbType.Int64}
+                    }
+                );
+                await Task.WhenAll(albums.Select(album =>
                 {
-                    cmd.Parameters["SingerId"].Value = singer.SingerId;
+                    cmd.Parameters["SingerId"].Value = album.SingerId;
+                    cmd.Parameters["AlbumId"].Value = album.AlbumId;
                     return cmd.ExecuteNonQueryAsync();
                 }));
 
-                Console.WriteLine("Deleted data.");
+                Console.WriteLine("Deleted individual rows in Albums.");
+
+                // Delete a range of rows from the Singers table where the column key is >=3 and <5.
+                cmd = connection.CreateDmlCommand(
+                   "DELETE FROM Singers WHERE SingerId >= 3 AND SingerId < 5");
+                int rowCount = await cmd.ExecuteNonQueryAsync();
+                Console.WriteLine($"{rowCount} row(s) deleted from Singers.");
+
+                // Delete remaining Singers rows, which will also delete the remaining Albums rows
+                // since it was defined with ON DELETE CASCADE.
+                cmd = connection.CreateDmlCommand(
+                   "DELETE FROM Singers WHERE true");
+                rowCount = await cmd.ExecuteNonQueryAsync();
+                Console.WriteLine($"{rowCount} row(s) deleted from Singers.");
```

Comment:
In some of the other languages, there is a way to call delete with a KetSet and a KeyRange. E.g. in Python the same sample looks like this:

```
    # Delete individual rows
    albums_to_delete = spanner.KeySet(
        keys=[[2, 1], [2, 3]])

    # Delete a range of rows where the column key is >=3 and <5
    singers_range = spanner.KeyRange(start_closed=[3], end_open=[5])
    singers_to_delete = spanner.KeySet(
        ranges=[singers_range])

    # Delete remaining Singers rows, which will also delete the remaining 
    # Albums rows since it was defined with ON DELETE CASCADE
    remaining_singers = spanner.KeySet(all_=True)

    with database.batch() as batch:
        batch.delete('Albums', albums_to_delete)
        batch.delete('Singers', singers_to_delete)
        batch.delete('Singers', remaining_singers)
```

I searched through the code to see if I can do something similar here but I don't think we can. `DbParameterCollection` doesn't have a concept like that so I don't think we can construct a delete command that will be equivalent to what Python is doing. This is why my implementation here resorts to a hand-written query.

But let me know if there's something I missed.

Reply:
Nope, we don't have anything like that, and we are sort of constraint by ADO.NET. It's possible we could add methods similar to those but it's not a usual pattern for ADO.NET providers as far as I know.

Reply:
Thanks, that's what I suspected. It's ok, I don't think we need to add the methods. If a user requests it, we can reconsider.